src/scheduler.cc

```diff
@@ -910,6 +910,31 @@ void Scheduler::tick(Locker& locker, int64 now) {
       // The process is already suppossed to preempt.
       // Check whether it is stuck.
       if (us_since_preemption <= MAX_RUN_WITHOUT_PREEMPTION_US) continue;
+      fprintf(stderr, "Potential dead-lock detected:\n");
+      fprintf(stderr, "  Process: %d\n", process->id());
+      if (process->current_bcp() != 0) {
+        uint8* last_bcp = process->current_bcp();
+        int bci = process->program()->absolute_bci_from_bcp(last_bcp);
+
+        const uint8* uuid = process->program()->id();
+        char uuid_buffer[37];
+        sprintf(uuid_buffer, "%08x-%04x-%04x-%04x-%04x%08x",
+            static_cast<int>(Utils::read_unaligned_uint32_be(uuid)),
+            static_cast<int>(Utils::read_unaligned_uint16_be(uuid + 4)),
+            static_cast<int>(Utils::read_unaligned_uint16_be(uuid + 6)),
+            static_cast<int>(Utils::read_unaligned_uint16_be(uuid + 8)),
+            static_cast<int>(Utils::read_unaligned_uint16_be(uuid + 10)),
+            static_cast<int>(Utils::read_unaligned_uint32_be(uuid + 12)));
+        fprintf(stderr, "  Program: %s\n", uuid_buffer);
+        fprintf(stderr, "  BCI: 0x%x\n", bci);
+
+        if (*last_bcp == Opcode::PRIMITIVE) {
```

Comment:
Can we check if the last bcp is within the program area?

Reply:
done.
I'm only checking if the bcp points into the bytecodes. Further accesses (like getting the primitive parameters) are not checked. But it's unlikely that we have a valid bytecode pointer that points to the edge of the bytecodes.